File: lib/trends/google-trends.ts
// Google Trends — no official API. Uses the public /trends/api/explore + widgetdata flow.
// Returns relative interest-over-time for one or more keywords + region.
// Failure modes: 429 rate-limit (most common), 200 with empty timelineData (silent), token expiry.

import { wrap, fetchJson, fmtNum } from './_helpers'
import type { TrendQuery, TrendResult, TrendDatapoint } from './types'

type ExploreWidget = { id: string; token: string; request: any }
type ExploreResponse = { widgets: ExploreWidget[] }
type TimelinePoint = { time: string; formattedTime?: string; formattedAxisTime?: string; value: number[]; formattedValue?: string[] }
type TimelineResponse = { default: { timelineData: TimelinePoint[] } }

export async function fetchGoogleTrends(query: TrendQuery): Promise<TrendResult> {
  return wrap('google-trends', query, async () => {
    const keywords = (query.keywords ?? []).slice(0, 5)
    if (keywords.length === 0) throw new Error('no keywords')
    const geo = (query.region ?? '').toUpperCase()
    const time = query.timeframe ?? 'today 12-m'

    const exploreReq = {
      comparisonItem: keywords.map(k => ({ keyword: k, geo: geo === 'WW' ? '' : geo, time })),
      category: 0,
      property: '',
    }
    const exploreUrl = `https://trends.google.com/trends/api/explore?hl=en-US&tz=0&req=${encodeURIComponent(JSON.stringify(exploreReq))}`
    const explore = await fetchJson<ExploreResponse>(exploreUrl)
    const timelineWidget = explore.widgets.find(w => w.id === 'TIMESERIES')
    if (!timelineWidget) throw new Error('no TIMESERIES widget')

    const wReq = { ...timelineWidget.request, requestOptions: { ...timelineWidget.request?.requestOptions, ...timelineWidget.request?.requestOptions } }
    const timelineUrl = `https://trends.google.com/trends/api/widgetdata/multiline?hl=en-US&tz=0&req=${encodeURIComponent(
      JSON.stringify(timelineWidget.request),
    )}&token=${encodeURIComponent(timelineWidget.token)}`
    const timeline = await fetchJson<TimelineResponse>(timelineUrl)
    const points = timeline.default?.timelineData ?? []
    if (points.length === 0) throw new Error('empty timeline')

    // Average per keyword across the window.
    const avgs = keywords.map((_, idx) => {
      const vals = points.map(p => p.value?.[idx] ?? 0).filter(v => !isNaN(v))
      return vals.reduce((s, v) => s + v, 0) / Math.max(1, vals.length)
    })
    const peakIdx = points.reduce((maxIdx, p, i, arr) => {
      const m = Math.max(...(p.value ?? []))
      const mc = Math.max(...(arr[maxIdx].value ?? []))
      return m > mc ? i : maxIdx
    }, 0)
    const peak = points[peakIdx]

    const datapoints: TrendDatapoint[] = points.map((p, i) => ({
      label: p.formattedTime ?? p.time ?? String(i),
      value: Math.max(...(p.value ?? [0])),
      meta: keywords.reduce((m, k, idx) => ({ ...m, [k]: p.value?.[idx] ?? 0 }), {}),
    }))

    const region = geo || 'WW'
    const summary = `关键词 ${keywords.join(' / ')} 在 ${region} 区域，${time} 时段平均热度 ${avgs.map(a => a.toFixed(0)).join(' / ')}（0-100 相对值）。`
    const lines = [
      `# Google Trends · ${keywords.join(' / ')} · ${region} · ${time}`,
      '',
      keywords.map((k, i) => `- **${k}** 平均热度 ${avgs[i].toFixed(1)}/100`).join('\n'),
      '',
      `**全期峰值**：${peak.formattedTime ?? peak.time}（最高值 ${Math.max(...(peak.value ?? [0]))}）`,
      '',
      `数据点 ${points.length} 个；首点 ${points[0]?.formattedTime ?? points[0]?.time}，尾点 ${points[points.length - 1]?.formattedTime ?? points[points.length - 1]?.time}。`,
    ]
    return { summary, digest: lines.join('\n'), datapoints }
  })
}

File: server/stratsquad/trends/google_trends.py
"""Google Trends — unofficial /trends/api/explore + widgetdata flow."""
from __future__ import annotations
import json
from urllib.parse import quote

from ..types import TrendQuery, TrendDatapoint, TrendResult
from .base import wrap, fetch_json


async def fetch_google_trends(query: TrendQuery) -> TrendResult:
    async def body() -> dict:
        keywords = (query.keywords or [])[:5]
        if not keywords:
            raise ValueError("no keywords")
        geo = (query.region or "").upper()
        time_range = query.timeframe or "today 12-m"

        explore_req = {
            "comparisonItem": [
                {"keyword": k, "geo": "" if geo == "WW" else geo, "time": time_range}
                for k in keywords
            ],
            "category": 0,
            "property": "",
        }
        explore_url = (
            "https://trends.google.com/trends/api/explore?hl=en-US&tz=0"
            f"&req={quote(json.dumps(explore_req))}"
        )
        explore = await fetch_json(explore_url)
        widget = next((w for w in explore["widgets"] if w["id"] == "TIMESERIES"), None)
        if not widget:
            raise ValueError("no TIMESERIES widget")

        timeline_url = (
            "https://trends.google.com/trends/api/widgetdata/multiline?hl=en-US&tz=0"
            f"&req={quote(json.dumps(widget['request']))}"
            f"&token={quote(widget['token'])}"
        )
        timeline = await fetch_json(timeline_url)
        points = (timeline.get("default") or {}).get("timelineData") or []
        if not points:
            raise ValueError("empty timeline")

        avgs = []
        for idx, _ in enumerate(keywords):
            vals = [p.get("value", [0])[idx] for p in points if p.get("value")]
            avgs.append(sum(vals) / max(1, len(vals)))

        peak = max(points, key=lambda p: max(p.get("value", [0]) or [0]))
        datapoints = [
            TrendDatapoint(
                label=p.get("formattedTime") or p.get("time") or str(i),
                value=float(max(p.get("value", [0]) or [0])),
                meta={k: p.get("value", [0])[ki] for ki, k in enumerate(keywords)},
            )
            for i, p in enumerate(points)
        ]

        region = geo or "WW"
        summary = f"关键词 {' / '.join(keywords)} 在 {region} 区域，{time_range} 时段平均热度 {' / '.join(f'{a:.0f}' for a in avgs)}（0-100 相对值）。"
        digest_lines = [
            f"# Google Trends · {' / '.join(keywords)} · {region} · {time_range}",
            "",
            "\n".join(f"- **{k}** 平均热度 {avgs[i]:.1f}/100" for i, k in enumerate(keywords)),
            "",
            f"**全期峰值**：{peak.get('formattedTime') or peak.get('time')}（最高值 {max(peak.get('value', [0]) or [0])}）",
            "",
            f"数据点 {len(points)} 个；首点 {points[0].get('formattedTime') or points[0].get('time')}，尾点 {points[-1].get('formattedTime') or points[-1].get('time')}。",
        ]
        return {"summary": summary, "digest": "\n".join(digest_lines), "datapoints": datapoints}

    return await wrap("google-trends", query, body)
